### PR TITLE
fix(common/core/web): OSK state-key management

### PR DESCRIPTION
**Pull request by jahorton**
Ran into this one while testing related aspects of #5451... turns out it was actually a bug on `master` _and_ `stable-14.0`.  (So, it'll need a 🍒-pick.)

## Base issue components & repros

Using either keymanweb.com (for stable) or the standard unminified test page:

1. Load a KMW-enabled page in your preferred Desktop browser.
    - The bugs are browser-independent.
2. Choose an input area (in order to display the OSK)
3. Press your keyboard's caps lock key:  it won't update.
4. Type an 'a' keypress - still no CAPS update.
5. Press the SHIFT key:  _now_ the OSK's caps key updates.

Then:

1. Refresh the page
2. Choose an input area (in order to display the OSK)
3. Press the OSK's caps-lock key:  it won't update
4. Press the OSK's 'a' key:  still no update
4. Press the OSK's SHIFT key:  still no update.  (So, broken OSK caps-lock.)
5. Press the keyboard's SHIFT key: still no caps-lock.  (The OSK should also drop out of SHIFT once the key is released.)

## More Description

Turns out I got related code "wrong" a bit during the 14.0 Core-Web refactor.  This PR will fix both issue components above:  the OSK will auto-update on any caps-lock keypress and will also toggle whenever the OSK's caps-lock key is pressed.
    - Note that a CAPS keypress after an OSK caps-press may not cause a change; physical keystrokes will immediately override the CAPS state with the physical keyboard's current caps-lock state.

## Notes

Remaining limitation:  if the active element loses focus with an active mnemonic keyboard and the caps-lock state is enabled, then regains it, the OSK currently tends to automatically select the SHIFT layer when redisplayed.  A physical SHIFT press & release will reset this.  (This is, at least, much better than auto-enabling the OSK's SHIFT for most keypresses whenever CAPS is on.)

## User Testing
@keymanapp/testers 

### Test suite 1: `desktop`

- [ ] Platform:  macOS / Safari

<details>
<summary>Instructions & Tests</summary>

Using the "Test unminified KeymanWeb" test page:

- [ ] TEST.STD.HARDWARE:
    <details>
    <summary><strong>Test sequence</strong></summary>
    
    1.  Ensure that your hardware keyboard's CAPS LOCK state is "off".
    2.  Select the "English - English" entry if not already selected.
        - The OSK's CAPS LOCK key should not be highlighted.
    3.  Press and release your hardware keyboard's CAPS LOCK key.
        - [ ] The OSK's copy of the key should become highlighted here.
    4.  Press and release your hardware keyboard's `A` key.
        - [ ] A lowercase `a` should result with no other effects.
    5.  Press and release your hardware keyboard's SHIFT.
        - [ ] The OSK's CAPS key should remain highlighted, while SHIFT highlights only while the physical SHIFT is pressed.
    6.  Press and release your hardware keyboard's CAPS LOCK key.
        - [ ] The OSK's CAPS key should no longer be highlighted.
    7.  Refresh the page.
    8. **Before displaying the OSK** or clicking anywhere within the page, activate your hardware keyboard's CAPS LOCK state.
    9. Display the OSK
        - [ ] The OSK's CAPS key should not be highlighted.  (It's a system limitation.)
    10. Press and release your hardware keyboard's `A` key.
        - [ ] The OSK's CAPS key should become highlighted.

    </details>

- [ ] TEST.STD.MIXED
    <details>
    <summary><strong>Test sequence</strong></summary>

    1.  Ensure that your hardware keyboard's CAPS LOCK state is "off".
    2.  Select the "English - English" entry if not already selected.
        - The OSK's CAPS LOCK key should not be highlighted.
    3.  Click the OSK's CAPS LOCK key.
        - [ ] It should become highlighted here.
    4.  Click your OSK's `A` key.
        - [ ] A lowercase `a` should result with no other effects.
    5.  Press and release your hardware keyboard's `A` key.
        - [ ] A lowercase `a` should result _and_ the OSK's CAPS LOCK key should lose its highlighting.
    6.  Press and release your hardware keyboard's CAPS LOCK key.
        - [ ] The OSK's CAPS key should become highlighted.
    7.  Click your OSK's CAPS key.
        - [ ] It should lose highlighting.
    8. Click your OSK's `A` key.
        - [ ] A lowercase `a` should result with no other effects.
    9. Press and release your hardware keyboard's `A` key.
        - [ ] A lowercase `a` should result _and_ the OSK's CAPS LOCK key should regain its highlighting.

    </details>

- [ ] TEST.MNEMONIC.HARDWARE:
    <details>
    <summary><strong>Test sequence</strong></summary>
    
    1.  Ensure that your hardware keyboard's CAPS LOCK state is "off".
    2.  Add the `sil_philippines` keyboard and select it.
        - The OSK's CAPS LOCK key should not be highlighted.
    3.  Press and release your hardware keyboard's CAPS LOCK key.
        - [ ] The OSK's copy of the key should become highlighted here.
    4.  Press and release your hardware keyboard's `A` key.
        - [ ] A capitalized `A` should result with no other effects.
    5.  Press and release your hardware keyboard's SHIFT.
        - [ ] The OSK's CAPS key should remain highlighted, while SHIFT highlights only while the physical SHIFT is pressed.
    6. Press, but do _not_ release your hardware keyboard's SHIFT.
        - [ ] The OSK's CAPS key should remain highlighted, while SHIFT also becomes highlighted.
    7. Press and release your hardware keyboard's `A` key.
        - [ ] A lowercased `a` should result with no other effects.
    8.  Release your hardware keyboard's SHIFT key.
        - [ ] The OSK's CAPS key should remain highlighted.

    </details>

- [ ] TEST.MNEMONIC.MIXED
    <details>
    <summary><strong>Test sequence</strong></summary>

    1.  Ensure that your hardware keyboard's CAPS LOCK state is "off".
    2.  Add the `sil_philippines` keyboard and select it.
        - The OSK's CAPS LOCK key should not be highlighted.
    3.  Click the OSK's CAPS LOCK key.
        - [ ] It should become highlighted here.
    4.  Click your OSK's `A` key.
        - [ ] A capitalized `A` should result with no other effects.
    5.  Press and release your hardware keyboard's `A` key.
        - [ ] A lowercase `a` should result _and_ the OSK's CAPS LOCK key should lose its highlighting.
    6.  Press and release your hardware keyboard's CAPS LOCK key.
        - [ ] The OSK's CAPS key should become highlighted.
    7.  Click your OSK's CAPS key.
        - [ ] It should lose highlighting.
    8. Click your OSK's `A` key.
        - [ ] A lowercase `a` should result with no other effects.
    9. Press and release your hardware keyboard's `A` key.
        - [ ] A capitalized `A` should result _and_ the OSK's CAPS LOCK key should regain its highlighting.

    </details>

</details>

### Test suite 2:  `mobile`

- [ ] Platform:  Chrome emulation / iOS

<details>
<summary>Instructions & Tests</summary>

Using the "Test unminified KeymanWeb" test page:

- [ ] TEST.MOBILE.NUM_LOCK:  Using the `khmer_angkor` keyboard, swap back and forth between the 'default' and the 'numeric' layers.
    - Expected result:  there should be no "stuck highlighting" effect on the default layer's numeric key.

</details>

### CODE REVIEW

File: common/core/web/input-processor/src/text/inputProcessor.ts
Comment (jahorton) [inline common/core/web/input-processor/src/text/inputProcessor.ts:79]: Keep in mind that with boolean operator [short-circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation), the `&&` is _**not**_ an associative operator - order matters.

We always want `doModifierPress` to run, as that triggers OSK state-key updates.  We simply don't want to instantly exit the function unless it's also caused by a physical keystroke.

File: web/source/osk/oskManager.ts
Comment (jahorton) [inline web/source/osk/oskManager.ts:318]: The reason physical caps-key presses didn't update the OSK's matching key.

File: web/source/osk/visualKeyboard.ts
Comment (jahorton) [inline web/source/osk/visualKeyboard.ts:903]: 1. This was the wrong way to do it, as evidenced by the fact that it doesn't actually trigger an OSK update.
2. It belongs within Core-Web.

File: common/core/web/keyboard-processor/src/keyboards/activeLayout.ts
Comment (jahorton) [inline common/core/web/keyboard-processor/src/keyboards/activeLayout.ts:238]: The replacement for the old, within-OSK approach.  This also correlates far more closely with what physical keystrokes do; they note the change in state, while the keyboard processor notices that change (hence `LmodifierChange`) and signals the need for an update.

File: common/core/web/input-processor/src/text/inputProcessor.ts
Comment (mcdurdin) [reply common/core/web/input-processor/src/text/inputProcessor.ts:79]: When we have side-effects like this, I like to structure the code to emphasize it:
```suggestion
      if(this.keyboardProcessor.doModifierPress(keyEvent, outputTarget, !fromOSK)) {
        if(!fromOSK) {
```

Makes it slightly less likely that a future maintainer won't accidentally swap them back...

File: common/core/web/keyboard-processor/src/keyboards/activeLayout.ts
Comment (mcdurdin) [inline common/core/web/keyboard-processor/src/keyboards/activeLayout.ts:233]:
```suggestion
        const bitmap = {
          'K_CAPS': text.Codes.stateBitmasks.CAPS,
          'K_NUMLOCK': text.Codes.stateBitmasks.NUM_LOCK,
          'K_SCROLL': text.Codes.stateBitmasks.SCROLL_LOCK
        };
        const bitmask = bitmap[Lkc.kName];
```

**Review by mcdurdin — APPROVED** [review]
LGTM pending user testing

File: common/core/web/input-processor/src/text/inputProcessor.ts
Comment (mcdurdin) [inline common/core/web/input-processor/src/text/inputProcessor.ts:84]: 👏

**Comment by MakaraSok** [comment]
Steps taken: (1) loaded PR, (2) navigated to .../web/source in Git Bash, (3) ran `./build.sh`, (4) open `http://localhost/keymanweb/testing/unminified.html#` after clearing cache in Chrome and FireFox. The results are the same: the <kbd>Caps</kbd> is highlighted in the OSK when pressing on the key on the OSK and the keyboard, yet small letters are output untill the Shift is triggered.

image

**Comment by jahorton** [comment]
> Steps taken: (1) loaded PR, (2) navigated to .../web/source in Git Bash, (3) ran `./build.sh`, (4) open `http://localhost/keymanweb/testing/unminified.html#` after clearing cache in Chrome and FireFox. The results are the same: the Caps is highlighted in the OSK when pressing on the key on the OSK and the keyboard, yet small letters are output untill the Shift is triggered.
> 
> image

When a Keyman keyboard does not define CAPS rules, no caps processing occurs.  I did have to do a double-check though; even for keyboards that don't explicitly define a rule for the base keystroke, this will hold - this is because we need to emulate the keystroke within KeymanWeb for use with predictive text, and our default output definitions do not account for CAPS state.

**Comment by mcdurdin** [comment]
> When a Keyman keyboard does not define CAPS rules, no caps processing occurs.

That's not quite true. A keyboard which uses characters for key rules will be impacted by caps lock, for example:

```
+ 'a' > 'little a'
+ 'A' > 'BIG A'
```

**Comment by jahorton** [comment]
> > When a Keyman keyboard does not define CAPS rules, no caps processing occurs.
> 
> That's not quite true. A keyboard which uses characters for key rules will be impacted by caps lock, for example:
> 
> ```
> + 'a' > 'little a'
> + 'A' > 'BIG A'
> ```

Wait... so the caps state should affect what mnemonic rule processing receives, rather than merely being part of the modifier metadata of the key event?

**Comment by jahorton** [comment]
> Gimme a specific keyboard to test if any. :)

`sil_philippines` is the first mnemonic keyboard I can think of.

Looks like KMW has been taking CAPS state into account, based on what I'm seeing compared to what Marc indicated above.

**Comment by mcdurdin** [comment]
> Wait... so the caps state should affect what mnemonic rule processing receives, rather than merely being part of the modifier metadata of the key event?

Warning: there may be some confusion here between mnemonic and positional. Even in a positional keyboard, using characters in the key part of the rule means we get US English default behaviour for those keys for Shift and Caps Lock. This corresponds to having the `ISVIRTUALKEY` flag = `0` in the modifier state. This functionality is active when `&mnemoniclayout` is `0`.

**Comment by jahorton** [comment]
> Warning: there may be some confusion here between mnemonic and positional. Even in a positional keyboard, using characters in the key part of the rule means we get US English default behaviour for those keys for Shift and Caps Lock. This corresponds to having the `ISVIRTUALKEY` flag = `0` in the modifier state. This functionality is active when `&mnemoniclayout` is `0`.

... those are different?  Up to this point, I thought those were two terms for the same type of keyboard.  I wasn't aware we supported positional but non-mnenomic keyboards.  Is this a legacy thing?

From an end-user perspective (or a keyboard designer perspective, if more appropriate), what's the functional difference between the two?

**Comment by mcdurdin** [comment]
Yeah, so this is a little confusing, agreed. We have four ways of matching key events:

1. Pure virtual key (positional); `ISVIRTUALKEY` bit is set, `VIRTUALCHARKEY` bit is clear:

   ```
   c matches SHIFT C01 (aka SHIFT K_A on US keyboard)
   c caps lock is ignored
   + [SHIFT K_A] >

   c C01 is an alias for K_A
   + [SHIFT C01] >

   c Explicitly specify Caps Lock with NCAPS and CAPS
   + [NCAPS SHIFT K_A] >
   + [CAPS K_A] >
   ```

2. US character code (positional); `ISVIRTUALKEY` and `VIRTUALCHARKEY` bits are clear:

    ```
    c matches NCAPS SHIFT C01 (aka NCAPS SHIFT K_A on US keyboard)
    c also matches CAPS C01.
    + 'A' >
    ```

3. Mnemonic character code. This matches the key+modifier state that generates this character on the user's base keyboard; `ISVIRTUALKEY` and `VIRTUALCHARKEY` bits are clear:

    ```
    c matches CAPS+D01 or SHIFT+D01 on US English
    c matches CAPS+C01 or SHIFT+C01 on French
    + 'Q' > 

    c matches SHIFT+K_LBRKT (D11) on US English
    + '{' > 
    ```

4. Mnemonic virtual character key; ``ISVIRTUALKEY` bit is clear (I think), `VIRTUALCHARKEY` bit is set:

    ```
    c matches SHIFT+c on user's base keyboard (i.e. SHIFT B03 on US)
    + [SHIFT 'c'] >

    c matches SHIFT+key that generates comma, i.e. SHIFT B08 on US
    c which means '<' on US English but on French (SHIFT B07) means '?'
    + [SHIFT ','] >  
    ```

I would like to try and make sure this is clear in the documentation -- it's kinda there in [virtual keys](https://help.keyman.com/developer/language/guide/virtual-keys) but not as explicitly documented as this.

Note that support for mnemonic layouts for base layouts other than US English, for web, is out of scope at present. But should maintain what we already have.

**Comment by mcdurdin** [comment]
Looks like mode #2 does not support Caps Lock (because when the relevant code was written in the KMW compiler, KMW did not support Caps Lock either...):

https://github.com/keymanapp/keyman/blob/64e7dca3806856034ce097b609e1d9332cf78bb2/windows/src/developer/TIKE/compile/CompileKeymanWeb.pas#L1241-L1246

This should be raised as a consistency issue...

**Comment by jahorton** [comment]
So... aside from (perhaps) a test round with the `sil_philippines` keyboard, anything holding this PR up?

**Comment by jahorton** [comment]
... and I found a decent reason.  If CAPS is active on a physical keyboard, then the OSK's caps is pressed, then a physical key typed, that's not auto-re-highlighting the CAPS key, even though the CAPS state from the physical keystroke is applied.

**Comment by jahorton** [comment]
Well... that was definitely messier than expected.  I get the feeling that some of the mnemonic processing stuff should be moved more completely into the keyboard processor at some point in the future.  But... since part of the goal is for this to be 🍒-pickable, those parts are best left 'til later.

**Comment by jahorton** [comment]
And... test suites enhanced.  Yikes, that's a "fun" one to spec up and format... but hey!  Nested collapsible sections!

Though, I didn't include the easy-copy thing I'd been doing.  Not like it was getting used, so I figured it'd be more readable to our tester this way.

And, to be explicit... @keymanapp/testers:  it's ready for a run.

**Comment by MakaraSok** [comment]
## User Testing
@keymanapp/testers 

### Test suite 1: `desktop`

- [x] Platform:  macOS / Safari

<details>
<summary>Instructions & Tests</summary>

Using the "Test unminified KeymanWeb" test page:

- [x] TEST.STD.HARDWARE: - **ALL PASSED**
    <details>
    <summary><strong>Test sequence</strong></summary>
    
    1.  Ensure that the hardware keyboard's CAPS LOCK state is "off".
    2.  Select the "English - English" entry if not already selected.
        - The OSK's CAPS LOCK key is not highlighted.
    3.  Press and release your hardware keyboard's CAPS LOCK key.
        - [x] **PASSED** - The OSK's copy of the key becomes highlighted.
    4.  Press and release your hardware keyboard's `A` key.
        - [x]  **PASSED** - A lowercase `a` results with no other effects.
    5.  Press and release your hardware keyboard's SHIFT.
        - [x] **PASSED** - The OSK's CAPS key remains highlighted, while SHIFT highlights only while the physical SHIFT is pressed.
    6.  Press and release your hardware keyboard's CAPS LOCK key.
        - [x] **PASSED** - The OSK's CAPS key is no longer highlighted.
    7.  Refresh the page.
    8. **Before displaying the OSK** or clicking anywhere within the page, activate your hardware keyboard's CAPS LOCK state.
    9. Display the OSK
        - [x] **PASSED** - The OSK's CAPS key is not highlighted.  _(It's a system limitation.)_
    10. Press and release your hardware keyboard's `A` key.
        - [x] **PASSED** - The OSK's CAPS key becomes highlighted.

    </details>

- [x] TEST.STD.MIXED - **ALL PASSED**
    <details>
    <summary><strong>Test sequence</strong></summary>

    1.  Ensure the hardware keyboard's CAPS LOCK state is "off".
    2.  Select the "English - English" entry if not already selected.
        - The OSK's CAPS LOCK key is not highlighted.
    3.  Click the OSK's CAPS LOCK key.
        - [x] **PASSED** - It becomes highlighted.
    4.  Click your OSK's `A` key.
        - [x] **PASSED** - A lowercase `a` results with no other effects.
    5.  Press and release your hardware keyboard's `A` key.
        - [x] **PASSED** - A lowercase `a` results _and_ the OSK's CAPS LOCK key loses its highlighting.
    6.  Press and release your hardware keyboard's CAPS LOCK key.
        - [x] **PASSED** - The OSK's CAPS key becomes highlighted.
    7.  Click your OSK's CAPS key.
        - [x] **PASSED** - It loses highlighting.
    8. Click your OSK's `A` key.
        - [x] **PASSED** - A lowercase `a` results with no other effects.
    9. Press and release your hardware keyboard's `A` key.
        - [x] **PASSED** - A lowercase `a` results _and_ the OSK's CAPS LOCK key should regain its highlighting.

    </details>

- [x] TEST.MNEMONIC.HARDWARE: - **ALL PASSED**
    <details>
    <summary><strong>Test sequence</strong></summary>
    
    1.  Ensure the hardware keyboard's CAPS LOCK state is "off".
    2.  Add the `sil_philippines` keyboard and select it.
        - The OSK's CAPS LOCK key is not highlighted.
    3.  Press and release your hardware keyboard's CAPS LOCK key.
        - [x] **PASSED** - The OSK's copy of the key does become highlighted here.
    4.  Press and release your hardware keyboard's `A` key.
        - [x] **PASSED** - A capitalized `A` does result with no other effects.
    5.  Press and release your hardware keyboard's SHIFT.
        - [x] **PASSED** - The OSK's CAPS key does remain highlighted, while SHIFT highlights only while the physical SHIFT is pressed.
    6. Press, but do _not_ release your hardware keyboard's SHIFT.
        - [x] **PASSED** - The OSK's CAPS key does remain highlighted, while SHIFT also becomes highlighted.
    7. Press and release your hardware keyboard's `A` key. 
        - [x] **PASSED** - A lowercased `a` does result with no other effects.
    8.  Release your hardware keyboard's SHIFT key.
        - [x] **PASSED** - The OSK's CAPS key should remain highlighted.

    </details>

- [x] TEST.MNEMONIC.MIXED - **ALL PASSED**
    <details>
    <summary><strong>Test sequence</strong></summary>

    1.  Ensure the hardware keyboard's CAPS LOCK state is "off".
    2.  Add the `sil_philippines` keyboard and select it.
        - The OSK's CAPS LOCK is indeed not highlighted.
    3.  Click the OSK's CAPS LOCK key.
        - [x] **PASSED** - It does become highlighted here.
    4.  Click your OSK's `A` key.
        - [x] **PASSED** - A capitalized `A` does result with no other effects.
    5.  Press and release your hardware keyboard's `A` key.
        - [x] **PASSED** - A lowercase `a` does result _and_ the OSK's CAPS LOCK key does lose its highlighting.
    6.  Press and release your hardware keyboard's CAPS LOCK key.
        - [x] **PASSED** - The OSK's CAPS key does become highlighted.
    7.  Click your OSK's CAPS key.
        - [x] **PASSED** - It does lose highlighting.
    8. Click your OSK's `A` key.
        - [x] **PASSED** - A lowercase `a` does result with no other effects.
    9. Press and release your hardware keyboard's `A` key.
        - [x] **PASSED** - A capitalized `A` does result _and_ the OSK's CAPS LOCK key does regain its highlighting.

    </details>

</details>

### Test suite 2:  `mobile`

- [x] Platform:  Chrome emulation / iOS

<details>
<summary>Instructions & Tests</summary>

Using the "Test unminified KeymanWeb" test page:

- [x] TEST.MOBILE.NUM_LOCK: **PASSED** - no "stuck highlighting" effects on the default layer's numeric key when swapping back and forth between the 'default' and the 'numeric' layers using the `khmer_angkor` keyboard.

</details>

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.89-alpha](https://keyman.com/downloads/releases/alpha/15.0.89)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.96-alpha](https://keyman.com/downloads/releases/alpha/15.0.96)

**Pull request merged**
